Installation issue: boost@1.70.0 on Fedora 35

### Steps to reproduce the issue

On an installation of fedora 35:
```shell
$ spack install boost@1.70.0 cxxstd=17 %gcc9.4.0
```
The outcome is a variety of error messages like:

```shell
 >> 16215    ./boost/thread/pthread/thread_data.hpp:60:5: error: missing binary operator before token "("
     16216       60 | #if PTHREAD_STACK_MIN > 0
...
```

### Information on your system

```shell
$ spack debug report
* **Spack:** 0.17.1-730-a4f0fbafbb
* **Python:** 3.10.1
* **Platform:** linux-fedora35-broadwell
* **Concretizer:** clingo
```

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/7818378/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/7818379/spack-build-env.txt)

@hainest 

As far as I can tell this is an issue related to recent changes in glibc.  Here's how they fixed it in other package managers: https://github.com/Ma27/nixpkgs/commit/87a628b1027363e8756c8bef7f6e4b71f190ee54

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/boost/package.py
+++ b/var/spack/repos/builtin/packages/boost/package.py
@@ -310,6 +310,10 @@ class Boost(Package):
           when="@1.77.0",
           working_dir="tools/build")
 
+    # Fix issues with PTHREAD_STACK_MIN not being a DEFINED constant in newer glibc
+    # See https:://github.com/spack/spack/issues/28273
+    patch("pthread-stack-min-fix.patch", when="@1.69.0:1.73.0")
+
     def patch(self):
         # Disable SSSE3 and AVX2 when using the NVIDIA compiler
         if self.spec.satisfies('%nvhpc'):

--- a/var/spack/repos/builtin/packages/boost/pthread-stack-min-fix.patch
+++ b/var/spack/repos/builtin/packages/boost/pthread-stack-min-fix.patch
@@ -1,0 +1,15 @@
+Taken from https://github.com/conan-io/conan-center-index/pull/361/files
+
+diff --git a/include/boost/thread/pthread/thread_data.hpp b/include/boost/thread/pthread/thread_data.hpp
+index aefbeb4..bc9b136 100644
+--- a/boost/thread/pthread/thread_data.hpp
++++ b/boost/thread/pthread/thread_data.hpp
+@@ -57,7 +57,7 @@ namespace boost
+ #else
+           std::size_t page_size = ::sysconf( _SC_PAGESIZE);
+ #endif
+-#if PTHREAD_STACK_MIN > 0
++#ifdef PTHREAD_STACK_MIN
+           if (size<PTHREAD_STACK_MIN) size=PTHREAD_STACK_MIN;
+ #endif
+           size = ((size+page_size-1)/page_size)*page_size;

